### PR TITLE
OHRM5X-1971: Wire app with OrangeHRM 5.x leave APIs

### PR DESCRIPTION
**Pull request by ShehanOHRM**
Update leave module

### CODE REVIEW

File: ios/Podfile.lock
Comment (RajithaKumara) [inline ios/Podfile.lock:1]: keep this lock file

File: src/screens/leave/LeaveDays.tsx [hunk 1]
```diff
@@ -26,15 +26,24 @@ import withTheme, {WithTheme} from 'lib/hoc/withTheme';
 import Divider from 'components/DefaultDivider';
 import {LeaveDaysRouteParams} from 'screens/leave/navigators';
 import LeaveDayListItem from 'screens/leave/components/LeaveDayListItem';
+import {fetchEmployeeLeaveRequest} from 'store/leave/leave-list/actions';
+import {connect} from 'react-redux';
+import {RootState} from 'store';
+import {selectEmployeeLeaveRequest} from '../../store/leave/leave-list/selectors';
```
Comment (RajithaKumara) [inline src/screens/leave/LeaveDays.tsx]:
```suggestion
import {selectEmployeeLeaveRequest} from 'store/leave/leave-list/selectors';
```

File: src/screens/leave/ApplyLeave.tsx [hunk 2]
```diff
@@ -31,6 +31,7 @@ import MainLayout from 'layouts/MainLayout';
 import withTheme, {WithTheme} from 'lib/hoc/withTheme';
 import {connect, ConnectedProps} from 'react-redux';
 import {RootState} from 'store';
+import {selectMyInfo} from '../../store/auth/selectors';
```
Comment (RajithaKumara) [inline src/screens/leave/ApplyLeave.tsx]:
```suggestion
import {selectMyInfo} from 'store/auth/selectors';
```

File: src/screens/leave/LeaveDetails.tsx [hunk 3]
```diff
@@ -265,10 +298,9 @@ class LeaveDetails extends React.Component<
                           ? {color: theme.typography.lightColor}
                           : {color: theme.typography.darkColor},
                       ]}>
-                      {employeeLeaveRequest?.leaveType.type
-                        ? employeeLeaveRequest?.leaveType.type +
-                          (employeeLeaveRequest?.leaveType.deleted ===
-                          LEAVE_TYPE_DELETED_YES
```
Comment (RajithaKumara) [inline src/screens/leave/LeaveDetails.tsx:271]: *

File: src/screens/leave/LeaveDetails.tsx [hunk 4]
```diff
@@ -205,7 +230,13 @@ class LeaveDetails extends React.Component<
                 paddingVertical: theme.spacing * 2,
                 paddingRight: theme.spacing * 4,
               }}>
-              <Avatar name={employeeLeaveRequest?.employeeName} />
+              <Avatar
+                name={
```
Comment (RajithaKumara) [inline src/screens/leave/LeaveDetails.tsx:212]: *

File: src/screens/leave/LeaveDetails.tsx [hunk 5]
```diff
@@ -149,39 +167,46 @@ class LeaveDetails extends React.Component<
...
-                  ) !== -1 ? (
-                    <View style={{paddingLeft: theme.spacing * 2}}>
-                      <Button
-                        title={'Approve'}
-                        primary
-                        onPress={this.onPressApproveLeave}
-                      />
-                    </View>
-                  ) : null}
+                  {employeeLeaveRequestDetails.allowedActions.map((item) => {
```
Comment (RajithaKumara) [inline src/screens/leave/LeaveDetails.tsx]: *

File: src/screens/leave/LeaveDetails.tsx [hunk 6]
```diff
@@ -149,39 +167,46 @@ class LeaveDetails extends React.Component<
...
-                  ) !== -1 ? (
-                    <Button
-                      title={'Cancel'}
-                      bordered
-                      primary
-                      onPress={this.onPressCancelLeave}
-                    />
-                  ) : null}
+                  {employeeLeaveRequestDetails.allowedActions.map((item) => {
+                    if (item.name === CANCEL) {
```
Comment (RajithaKumara) [inline src/screens/leave/LeaveDetails.tsx]: *

File: src/screens/leave/LeaveDetails.tsx [hunk 7]
```diff
@@ -69,17 +78,17 @@ class LeaveDetails extends React.Component<
...
+      this.props.fetchEmployeeLeaveRequestDetails(leaveRequest.id);
+      this.props.fetchEmployeeLeaveRequest(leaveRequest.id);
+      this.props.fetchLeaveComment(leaveRequest.id);
     }
   }
 
   onRefresh = () => {
     const {leaveRequest} = this.props.route.params;
-    this.props.fetchEmployeeLeaveRequest(leaveRequest.leaveRequestId);
+    this.props.fetchEmployeeLeaveRequest(leaveRequest.id);
```
Comment (RajithaKumara) [inline src/screens/leave/LeaveDetails.tsx]:
```suggestion
    this.props.fetchEmployeeLeaveRequestDetails(leaveRequest.id);
    this.props.fetchEmployeeLeaveRequest(leaveRequest.id);
    this.props.fetchLeaveComment(leaveRequest.id);
```
:thinking:

File: src/lib/helpers/leave.ts [hunk 8]
```diff
@@ -45,8 +45,8 @@ export const LEAVE_TYPE_COLORS = [
 
 const LEAVE_STATUS_MAP = {
```
Comment (RajithaKumara) [inline src/lib/helpers/leave.ts:48]: TODO:: will remove

File: src/lib/helpers/leave.ts [hunk 9]
```diff
@@ -66,6 +66,15 @@ const assignColorsToLeaveTypes = <T extends Data>(data: T[]): T[] => {
   return newDataArray;
 };
 
+const assignLeaveBalance = <T extends Data>(data: T[], data1: T[]): T[] => {
```
Comment (RajithaKumara) [inline src/lib/helpers/leave.ts]: seems not using

File: src/screens/leave/LeaveDetails.tsx [hunk 10]
```diff
@@ -99,45 +108,54 @@ class LeaveDetails extends React.Component<
...
   };
 
   render() {
-    const {theme, employeeLeaveRequest} = this.props;
+    const {theme, employeeLeaveRequestDetails, employeeLeaveComment} =
+      this.props;
     const {action} = this.state;
-    const leaveTypeColor = employeeLeaveRequest?.leaveType.color;
+
+    const leaveTypeColor = 'green';
```
Comment (RajithaKumara) [inline src/screens/leave/LeaveDetails.tsx]: ??

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: package.json [hunk 11]
```diff
@@ -21,14 +22,12 @@
     "@react-navigation/drawer": "~6.5.7",
     "@react-navigation/native": "~6.0.0",
     "@react-navigation/stack": "~6.3.11",
-    "connected-react-router": "~6.0.0",
-    "deprecated-react-native-prop-types": "~4.0.0",
-    "eslint-plugin-prettier": "^4.2.1",
```
Comment (RajithaKumara) [inline package.json:26]: @ShehanOHRM I removed these. please double check

File: package.json [hunk 12]
```diff
@@ -42,7 +41,6 @@
     "redux-saga": "~1.2.2",
     "reselect": "~4.0.0",
     "semver": "~7.3.4",
-    "ts-jest": "^29.0.5",
```
Comment (RajithaKumara) [inline package.json:45]: @ShehanOHRM I removed this also. please double check

File: package.json [hunk 13]
```diff
@@ -21,14 +22,12 @@
     "@react-navigation/drawer": "~6.5.7",
     "@react-navigation/native": "~6.0.0",
     "@react-navigation/stack": "~6.3.11",
-    "connected-react-router": "~6.0.0",
-    "deprecated-react-native-prop-types": "~4.0.0",
-    "eslint-plugin-prettier": "^4.2.1",
+    "lodash": "~4.17.21",
```
Comment (RajithaKumara) [inline package.json:25]: added to get debounce

File: src/screens/leave/LeaveDetails.tsx [hunk 14]
```diff
@@ -149,39 +152,46 @@ class LeaveDetails extends React.Component<
...
-                  ) !== -1 ? (
-                    <View style={{paddingLeft: theme.spacing * 2}}>
-                      <Button
-                        title={'Approve'}
-                        primary
-                        onPress={this.onPressApproveLeave}
-                      />
-                    </View>
-                  ) : null}
+                  {employeeLeaveRequest.allowedActions.map((item) => {
```
Comment (RajithaKumara) [inline src/screens/leave/LeaveDetails.tsx]: TODO:: visually button order seems wrong

File: src/store/leave/assign-leave/sagas.ts [hunk 15]
```diff
@@ -123,15 +136,16 @@ function* fetchSubordinateLeaveEntitlements(
...
-      yield put(fetchSubordinateLeaveEntitlementsFinished([]));
-    } else if (response.error[0] === 'No Leave Types Defined.') {
-      yield put(
-        setErrorMessage(
-          'There Are No Leave Types Defined, Please Contact Your System Administrator.',
-        ),
-      );
-    } else {
+    }
+    // TODO::  handle errors
```
Comment (RajithaKumara) [inline src/store/leave/assign-leave/sagas.ts:140]: TODO

File: src/store/saga-effects/api.ts [hunk 16]
```diff
@@ -34,11 +34,17 @@ import {
...
-import {getExpiredAt} from 'store/auth/helper';
 import {AuthParams} from 'store/storage/types';
 import {logout, fetchNewAuthTokenFinished} from 'store/auth/actions';
 import {AuthenticationError} from 'services/errors/authentication';
 
+export interface ApiResponse<Data, Meta = {}> {
+  data: Data;
+  meta: Meta;
+  // eslint-disable-next-line no-undef
+  getResponse: () => Response;
```
Comment (RajithaKumara) [inline src/store/saga-effects/api.ts:45]: TODO:: add error attributes

File: src/store/saga-effects/api.ts [hunk 17]
```diff
@@ -70,7 +76,15 @@ export function* apiCall<Fn extends (...args: any[]) => any>(
         authParams.instanceUrl,
         authParams.refreshToken,
       );
-      const data = yield call([response, response.json]);
+      const data: {
+        access_token: string;
+        refresh_token: string;
+        token_type: string;
+        expires_in: number;
+        scope: string; // TODO
```
Comment (RajithaKumara) [inline src/store/saga-effects/api.ts:84]: TODO

**Review by RajithaKumara — APPROVED** [review]
Some feedback added https://github.com/ShehanOHRM/orangehrm-os-mobile/pull/2#pullrequestreview-1368256804

File: src/screens/leave/LeaveDetails.tsx (hunk 10 above)
Comment (ShehanOHRM) [reply src/screens/leave/LeaveDetails.tsx]: leaveType.color, not comming data in latest api

**Pull request merged**